lightning/src/chain/package.rs

```diff
@@ -0,0 +1,1037 @@
+	preimage,
+	amount
+});
+
+/// A struct to describe the channel output on the funding transaction.
+///
+/// witnessScript is used as part of the witness redeeming the funding utxo.
+#[derive(Clone, PartialEq)]
+pub(crate) struct HolderFundingOutput {
+	funding_redeemscript: Script,
+}
+
+impl HolderFundingOutput {
+	pub(crate) fn build(funding_redeemscript: Script) -> Self {
+		HolderFundingOutput {
+			funding_redeemscript,
+		}
+	}
+}
+
+impl_writeable!(HolderFundingOutput, 0, {
+	funding_redeemscript
+});
+
+/// A wrapper encapsulating all in-protocol differing outputs types.
+///
+/// The generic API offers access to outputs attributes (amount, weight, output_type) or allow
+/// transformation such as finalizing an input claiming the output.
+#[derive(Clone, PartialEq)]
+pub(crate) enum PackageSolvingData {
+	RevokedOutput(RevokedOutput),
+	RevokedHTLCOutput(RevokedHTLCOutput),
+	CounterpartyOfferedHTLCOutput(CounterpartyOfferedHTLCOutput),
+	CounterpartyReceivedHTLCOutput(CounterpartyReceivedHTLCOutput),
+	HolderHTLCOutput(HolderHTLCOutput),
+	HolderFundingOutput(HolderFundingOutput),
+}
+
+impl PackageSolvingData {
+	fn amount(&self) -> u64 {
+		let amt = match self {
+			PackageSolvingData::RevokedOutput(ref outp) => { outp.amount },
+			PackageSolvingData::RevokedHTLCOutput(ref outp) => { outp.amount },
+			PackageSolvingData::CounterpartyOfferedHTLCOutput(ref outp) => { outp.htlc.amount_msat / 1000 },
+			PackageSolvingData::CounterpartyReceivedHTLCOutput(ref outp) => { outp.htlc.amount_msat / 1000 },
+			// Note: Currently, amounts of holder outputs spending witnesses aren't used
+			// as we can't malleate spending package to increase their feerate. This
+			// should change with the remaining anchor output patchset.
+			PackageSolvingData::HolderHTLCOutput(..) => { 0 },
+			PackageSolvingData::HolderFundingOutput(..) => { 0 },
+		};
+		amt
+	}
+	fn weight(&self) -> usize {
+		let weight = match self {
+			PackageSolvingData::RevokedOutput(ref outp) => { outp.weight as usize },
+			PackageSolvingData::RevokedHTLCOutput(ref outp) => { outp.weight as usize },
+			PackageSolvingData::CounterpartyOfferedHTLCOutput(..) => { WEIGHT_OFFERED_HTLC as usize },
+			PackageSolvingData::CounterpartyReceivedHTLCOutput(..) => { WEIGHT_RECEIVED_HTLC as usize },
+			// Note: Currently, weights of holder outputs spending witnesses aren't used
+			// as we can't malleate spending package to increase their feerate. This
+			// should change with the remaining anchor output patchset.
+			PackageSolvingData::HolderHTLCOutput(..) => { debug_assert!(false); 0 },
+			PackageSolvingData::HolderFundingOutput(..) => { debug_assert!(false); 0 },
+		};
+		weight
+	}
+	fn output_type_eq(&self, input: &PackageSolvingData) -> bool {
+		match self {
+			PackageSolvingData::RevokedOutput(..) => {
+				match input {
+					PackageSolvingData::RevokedHTLCOutput(..) => { return true },
+					PackageSolvingData::RevokedOutput(..) => { return true },
```

Comment:
If these variants only differ by one field but are considered "equal", would it be simpler to make that particular field an enum and collapse these two variants into one variant rather than repeat the six other fields across each?

Reply:
I was anticipating future implementation of disaggregation-after-timelock to solve this transaction pinning : https://github.com/lightningnetwork/lightning-rfc/pull/803. If this mitigation is implemented they won't be considered "equal" anymore.